test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,180 @@
+  }
+
+  if (!ECDSA_sign(0, hash, hash_len, ops->out_, &ops->out_len_, ec_key)) {
+    EC_KEY_free(ec_key);
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  EC_KEY_free(ec_key);
+  // Tell SSL socket that the operation is ready to be called again.
+  ops->delayed_op();
+
+  return ssl_private_key_retry;
+}
+
+static ssl_private_key_result_t privateKeyDecrypt(SSL* ssl, uint8_t* out, size_t* out_len,
+                                                  size_t max_out, const uint8_t* in,
+                                                  size_t in_len) {
+  (void)ssl;
+  (void)out;
+  (void)out_len;
+  (void)max_out;
+  (void)in;
+  (void)in_len;
+
+  return ssl_private_key_failure;
+}
+
+static ssl_private_key_result_t privateKeyComplete(SSL* ssl, uint8_t* out, size_t* out_len,
+                                                   size_t max_out) {
+  EcdsaPrivateKeyConnection* ops = static_cast<EcdsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index));
+
+  if (!ops->finished_) {
+    // The operation didn't finish yet, retry.
+    return ssl_private_key_retry;
+  }
+
+  if (ops->test_options_.async_method_error_) {
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  if (ops->out_len_ > max_out) {
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  memcpy(out, ops->out_, ops->out_len_);
+  *out_len = ops->out_len_;
+
+  OPENSSL_free(ops->out_);
+
+  return ssl_private_key_success;
+}
+
+Ssl::BoringSslPrivateKeyMethodSharedPtr
+EcdsaPrivateKeyMethodProvider::getBoringSslPrivateKeyMethod() {
+  return method_;
+}
+
+EcdsaPrivateKeyConnection::EcdsaPrivateKeyConnection(
+    SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher,
+    bssl::UniquePtr<EVP_PKEY> pkey, EcdsaPrivateKeyConnectionTestOptions& test_options)
+    : test_options_(test_options), cb_(cb), dispatcher_(dispatcher), pkey_(move(pkey)) {
+  SSL_set_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index, this);
+}
+
+Ssl::PrivateKeyConnectionPtr EcdsaPrivateKeyMethodProvider::getPrivateKeyConnection(
+    SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher) {
+  bssl::UniquePtr<BIO> bio(
+      BIO_new_mem_buf(const_cast<char*>(private_key_.data()), private_key_.size()));
+  bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
+  if (pkey == nullptr) {
+    return nullptr;
+  }
+
+  return std::make_unique<EcdsaPrivateKeyConnection>(ssl, cb, dispatcher, move(pkey),
+                                                     test_options_);
+}
+
+EcdsaPrivateKeyMethodProvider::EcdsaPrivateKeyMethodProvider(
+    const ProtobufWkt::Struct& config,
+    Server::Configuration::TransportSocketFactoryContext& factory_context) {
+
+  std::string private_key_path;
+
+  if (EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index == -1) {
```

Comment:
(This assumes that no two threads call this function at the same time. Could well be true, just highlighting the assumption.)

Reply:
In my understanding this is true, because the providers are initialized only from the main thread as the TLS context configuration is processed. If there are several providers in the TLS context configuration (two certificates) the initialization is done sequentially.